org.lflang/src/org/lflang/federated/FedASTUtils.java

```diff
@@ -74,25 +74,203 @@
+        LfFactory factory = LfFactory.eINSTANCE;
+
+        Action action = factory.createAction();
+        // Name the newly created action; set its delay and type.
+        action.setName(ASTUtils.getUniqueIdentifier(top, "networkMessage"));
+        if (serializer == SupportedSerializers.NATIVE) {
+            action.setType(type);
+        } else {
+            Type action_type = factory.createType();
+            action_type.setId(networkBufferType);
+            action.setType(action_type);
+        }
+        
+        // The connection is 'physical' if it uses the ~> notation.
+        if (connection.isPhysical()) {
+            action.setOrigin(ActionOrigin.PHYSICAL);
+            // Messages sent on physical connections do not
+            // carry a timestamp, or a delay. The delay
+            // provided using after is enforced by setting
+            // the minDelay.
+            if (connection.getDelay() != null) {
+                action.setMinDelay(factory.createValue());
+                action.getMinDelay().setTime(factory.createTime());
+                action.getMinDelay().getTime().setInterval(connection.getDelay().getInterval());
+                action.getMinDelay().getTime().setUnit(connection.getDelay().getUnit());
+            }
+        } else {
+            action.setOrigin(ActionOrigin.LOGICAL);
+        }
+        
+        return action;
     }
 
     /**
-     * Add a network control reaction for a given input port "portRef" to the
-     * reaction queue of the federated reactor. This reaction will block for
+     * Add a network receiver reaction for a given input port 'destination' to
+     * destination's parent reactor. This reaction will react to a generated
+     * 'networkAction' (triggered asynchronously, e.g., by federate.c). This
+     * 'networkAction' will contain the actual message that is sent by the sender
+     * in 'action->value'. This value is forwarded to 'destination' in the network
+     * receiver reaction.
+     * 
+     * @note: Used in federated execution
+     * 
+     * @param networkAction The network action (also, @see createNetworkAction)
+     * @param source The source port instance.
+     * @param destination The destination port instance.
+     * @param connection The network connection.
+     * @param sourceFederate The source federate.
+     * @param destinationFederate The destination federate.
+     * @param rightBankIndex The right bank index or -1 if the right reactor is not in a bank.
+     * @param rightChannelIndex The right channel index or -1 if the right port is not a multiport.
+     * @param generator The GeneratorBase instance used to perform some target-specific actions
+     * @param coordination One of CoordinationType.DECENTRALIZED or CoordinationType.CENTRALIZED.
+     * @param serializer The serializer used on the connection
+     */
+    private static void addNetworkReceiverReaction(
+            Action networkAction,
```

Comment:
The number of arguments of this function (as well as it's length) suggests that it's doing too many things at once. Without having carefully looked at what this function does exactly, I bet this can be factored out in two or three smaller functions, each of which has only a few arguments.

Reply:
That's a great a suggestion. I will take a stab at this refactoring in a day or two.